File: common/src/client/java/juuxel/adorn/client/renderer/ShelfRenderer.java
package juuxel.adorn.client.renderer;

import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.math.Axis;
import juuxel.adorn.block.ShelfBlock;
import juuxel.adorn.block.entity.ShelfBlockEntity;
import net.minecraft.client.renderer.SubmitNodeCollector;
import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.client.renderer.feature.ModelFeatureRenderer;
import net.minecraft.client.renderer.item.ItemModelResolver;
import net.minecraft.client.renderer.item.ItemStackRenderState;
import net.minecraft.client.renderer.state.level.CameraRenderState;
import net.minecraft.client.renderer.texture.OverlayTexture;
import net.minecraft.core.Direction;
import net.minecraft.world.item.ItemDisplayContext;
import net.minecraft.world.phys.Vec3;
import org.jspecify.annotations.Nullable;

public final class ShelfRenderer implements BlockEntityRenderer<ShelfBlockEntity, ShelfRenderState> {
    private static final float ITEM_SCALE = 0.5f;
    private static final float ITEM_1_Y_ROT = 10f;
    private static final float ITEM_2_Y_ROT = -17f;

    private final ItemModelResolver itemModelManager;

    public ShelfRenderer(BlockEntityRendererProvider.Context context) {
        this.itemModelManager = context.itemModelResolver();
    }

    @Override
    public ShelfRenderState createRenderState() {
        return new ShelfRenderState();
    }

    @Override
    public void extractRenderState(ShelfBlockEntity blockEntity, ShelfRenderState state, float tickProgress, Vec3 cameraPos, ModelFeatureRenderer.@Nullable CrumblingOverlay crumblingOverlay) {
        BlockEntityRenderer.super.extractRenderState(blockEntity, state, tickProgress, cameraPos, crumblingOverlay);

        int seed = (int) blockEntity.getBlockPos().asLong();

        var first = new ItemStackRenderState();
        itemModelManager.updateForTopItem(first, blockEntity.getItem(0), ItemDisplayContext.FIXED, blockEntity.getLevel(), null, seed);
        state.firstItem = first;

        var second = new ItemStackRenderState();
        itemModelManager.updateForTopItem(second, blockEntity.getItem(1), ItemDisplayContext.FIXED, blockEntity.getLevel(), null, seed + 1);
        state.secondItem = second;
    }

    @Override
    public void submit(ShelfRenderState state, PoseStack matrices, SubmitNodeCollector queue, CameraRenderState cameraState) {
        Direction facing = state.blockState.getValue(ShelfBlock.FACING);

        // For first item
        double tx1 = switch (facing) {
            case SOUTH, WEST -> 12 / 16.0;
            default -> 4 / 16.0;
        };
        double tz1 = switch (facing) {
            case NORTH, WEST -> 12 / 16.0;
            default -> 4 / 16.0;
        };

        // For second item
        double tx2 = switch (facing) {
            case NORTH, WEST -> 12 / 16.0;
            default -> 4 / 16.0;
        };
        double tz2 = switch (facing) {
            case NORTH, EAST -> 12 / 16.0;
            default -> 4 / 16.0;
        };

        matrices.pushPose();
        matrices.translate(tx1, 9.6 / 16.0, tz1);
        matrices.scale(ITEM_SCALE, ITEM_SCALE, ITEM_SCALE);
        matrices.mulPose(Axis.YP.rotationDegrees(ITEM_1_Y_ROT + 180 - facing.toYRot()));
        state.firstItem.submit(matrices, queue, state.lightCoords, OverlayTexture.NO_OVERLAY, 0);
        matrices.popPose();

        matrices.pushPose();
        matrices.translate(tx2, 9.6 / 16.0, tz2);
        matrices.scale(ITEM_SCALE, ITEM_SCALE, ITEM_SCALE);
        matrices.mulPose(Axis.YP.rotationDegrees(ITEM_2_Y_ROT + 180 - facing.toYRot()));
        state.secondItem.submit(matrices, queue, state.lightCoords, OverlayTexture.NO_OVERLAY, 0);
        matrices.popPose();
    }
}

File: src/main/kotlin/juuxel/adorn/block/renderer/ShelfRenderer.kt
package juuxel.adorn.block.renderer

import juuxel.adorn.block.ShelfBlock
import juuxel.adorn.block.entity.ShelfBlockEntity
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.MinecraftClient
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.render.block.entity.BlockEntityRenderDispatcher
import net.minecraft.client.render.block.entity.BlockEntityRenderer
import net.minecraft.client.render.model.json.ModelTransformation
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.client.util.math.Vector3f
import net.minecraft.util.math.Direction

@Environment(EnvType.CLIENT)
class ShelfRenderer(dispatcher: BlockEntityRenderDispatcher) : BlockEntityRenderer<ShelfBlockEntity>(dispatcher) {
    override fun render(
        be: ShelfBlockEntity, tickDelta: Float,
        matrices: MatrixStack, vertexConsumerProvider: VertexConsumerProvider, light: Int, overlay: Int
    ) {
        val facing = be.cachedState[ShelfBlock.FACING]

        // For first item
        val tx1 = when (facing) {
            Direction.SOUTH, Direction.WEST -> 12 / 16.0
            else -> 4 / 16.0
        }
        val tz1 = when (facing) {
            Direction.NORTH, Direction.WEST -> 12 / 16.0
            else -> 4 / 16.0
        }

        // For second item
        val tx2 = when (facing) {
            Direction.NORTH, Direction.WEST -> 12 / 16.0
            else -> 4 / 16.0
        }
        val tz2 = when (facing) {
            Direction.NORTH, Direction.EAST -> 12 / 16.0
            else -> 4 / 16.0
        }

        val itemRenderer = MinecraftClient.getInstance().itemRenderer

        matrices.push()
        matrices.translate(tx1, 9.6 / 16.0, tz1)
        matrices.scale(ITEM_SCALE, ITEM_SCALE, ITEM_SCALE)
        matrices.multiply(Vector3f.POSITIVE_Y.getDegreesQuaternion(ITEM_1_Y_ROT + facing.asRotation()))
        // GlStateManager.rotatef(ITEM_X_ROT, 1f, 0f, 0f)
        itemRenderer.renderItem(be.getInvStack(0), ModelTransformation.Mode.FIXED, light, overlay, matrices, vertexConsumerProvider)
        matrices.pop()

        matrices.push()
        matrices.translate(tx2, 9.6 / 16.0, tz2)
        matrices.scale(ITEM_SCALE, ITEM_SCALE, ITEM_SCALE)
        matrices.multiply(Vector3f.POSITIVE_Y.getDegreesQuaternion(ITEM_2_Y_ROT + facing.asRotation()))
        // GlStateManager.rotatef(ITEM_X_ROT, 1f, 0f, 0f)
        itemRenderer.renderItem(be.getInvStack(1), ModelTransformation.Mode.FIXED, light, overlay, matrices, vertexConsumerProvider)
        matrices.pop()
    }

    companion object {
        private const val ITEM_SCALE: Float = 0.5f
        // private const val ITEM_X_ROT: Float = 90f
        private const val ITEM_1_Y_ROT: Float = 10f
        private const val ITEM_2_Y_ROT: Float = -17f
    }
}
